content/courses/intro-to-embedded-sys-prog/chapters/low_level_programming.rst

```diff
@@ -334,86 +341,151 @@ the language evolves, just in case somebody does use them.
+     Post => MemCopy'Result = Destination;
+   --  Copies Length bytes from the object designated by Source to the object
+   --  designated by Destination. 
+
+The three aspects that do the importing are specified after the reserved 
+word :ada:`with` but can be ignored for this discussion. We'll talk 
+about them later. The preconditions make explicit the otherwise implicit 
+requirements for the arguments passed to memcpy, and the postcondition 
+specifies the expected result returned from a successful call. Neither 
+the preconditions nor the postconditions are required for importing 
+external entities but they are good "guard-rails" for using those 
+entities. If we call it incorrectly the precondition will inform us, and 
+likewise, if we misunderstand the result the postcondition will let us 
+know (at least to the extent that the return value does that). 
+
+For a sample call to our imported routine, imagine that we have a 
+procedure that copies the bytes of a String parameter into a Buffer 
+parameter, which is just a contiguous array of bytes. We need to tell 
+MemCopy the addresses of the arguments passed so we apply the attribute 
+accordingly: 
 
-When we declare the :ada:`Buffer_Overlay` object we also specify its
-address |mdash| strictly speaking its starting address |mdash| which,
-not surprisingly, is the address of the :ada:`Buffer` component at index
-:ada:`Start`. Applying the :ada:`Address` attribute gives us that
-location.
+.. code-block:: ada
 
-There are other characteristics we might want to query too. For example,
-we might want to ask the compiler what alignment it chose for a
-given object (or type, for all such objects). Rather than describe all
-the possibilities, we can just say that all the representation
-characteristics that can be specified can also be queried. We cover
-specifying representation characteristics next, so just assume the
-corresponding queries are available.
+   procedure Put (This : in out Buffer; Start : Index; Value : String) is
+      Result : System.Address with Unreferenced;
+   begin
+      Result := MemCopy (Destination => This (Start)'Address,
+                         Source      => Value'Address,
+                         Length      => Value'Length);
+   end Put;
+   
+The order of the address parameters is easily confused so we use the 
+named association format for specifying the actual parameters in the 
+call. 
+
+Although we assign Result we don't otherwise use it, so we tell the
+compiler this is not a mistake via the :ada:`Unreferenced` aspect. And if we
+do turn around and reference it the compiler will complain, as it should.
+
+(We don't show the preconditions for Put, but they would have specified 
+that Start must be a valid index into This particular Buffer, and that 
+there must be room in the Buffer argument for the number of bytes in 
+Value when starting at the Start index, so that we don't copy past the 
+end of the Buffer argument.) 
```

Comment:
```suggestion
(We don't show the preconditions for :ada:`Put`, but they would have specified 
that :ada:`Start` must be a valid index into this particular buffer, and that 
there must be room in the :ada:`Buffer` argument for the number of bytes in 
:ada:`Value` when starting at the :ada:`Start` index, so that we don't copy past the 
end of the :ada:`Buffer` argument.) 
```